niftypet/nipet/prj/src/recon.cu

```diff
@@ -2,14 +2,210 @@
 CUDA C extention for Python
 Provides functionality for PET image reconstruction.
 
-author: Pawel Markiewicz
-Copyrights: 2018
+Copyrights:
+2018-2020 Pawel Markiewicz
+2020 Casper da Costa-Luis
 ------------------------------------------------------------------------*/
 #include "recon.h"
+#include <assert.h>
 
 //number of threads used for element-wise GPU calculations
 #define NTHRDS 1024
+#define FLOAT_WITHIN_EPS(x) (-0.000001f < x && x < 0.000001f)
+
+/// z: how many Z-slices to add
+__global__ void pad(float *dst, float *src, const int z) {
+  int i = threadIdx.x + blockDim.x * blockIdx.x;
+  if (i >= SZ_IMX)
+    return;
+  int j = threadIdx.y + blockDim.y * blockIdx.y;
+  if (j >= SZ_IMY)
+    return;
+  src += i * SZ_IMY * SZ_IMZ + j * SZ_IMZ;
+  dst += i * SZ_IMY * (SZ_IMZ + z) + j * (SZ_IMZ + z);
+  for (int k = 0; k < SZ_IMZ; ++k)
+    dst[k] = src[k];
+}
+void d_pad(float *dst, float *src, const int z = COLUMNS_BLOCKDIM_X - SZ_IMZ % COLUMNS_BLOCKDIM_X) {
+  HANDLE_ERROR(cudaMemset(dst, 0, SZ_IMX * SZ_IMY * (SZ_IMZ + z) * sizeof(float)));
+  dim3 BpG((SZ_IMX + NTHRDS / 32 - 1) / (NTHRDS / 32), (SZ_IMY + 31) / 32);
+  dim3 TpB(NTHRDS / 32, 32);
+  pad<<<BpG, TpB>>>(dst, src, z);
+}
```

Comment:
NB:

- This is requires as z-padding occurs at regular intervals in `dst[]`, it's not as simple as padding the end of the array. Look at this array pointer offset:

```cpp
dst += i * SZ_IMY * (SZ_IMZ + z) + j * (SZ_IMZ + z);
```

- alternatively using (vis pre 6df9dfd)

```cpp
for (size_t i = 0; i < SZ_IMX; ++i)
  for (size_t j = 0; j < SZ_IMY; ++j)
    cudaMemcpy(&dst[i * SZ_IMY * (SZ_IMZ + z) + j * (SZ_IMZ + z)],
               &src[i * SZ_IMY * SZ_IMZ + j * SZ_IMZ],
               SZ_IMZ * sizeof(float), cudaMemcpyDeviceToDevice);
```

is very slow. It's quicker to use a CUDA kernel rather than calling `cudaMemcpy` 320x320 times.

- alternatively, "fixing" `d_conv` to work on `z=127` is too painful for me at the moment.